Add patchy pair potential

cbkerr opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
Clean and publish code to simulate patchy particles, including janus particles, based on that used in [Phase behavior of Janus colloids determined by sedimentation equilibrium.
](https://pubs.rsc.org/en/content/articlelanding/2014/sm/c3sm53136h)

Using templates, the code can "modulate" any isotropic pair potential by an "envelope" that is created based on an angle dependence factor.

## TODOS remaining

* [x] Add support for patches not aligned with particle orientation.
* [x] Add tests in a new file
* [x] Check casting, line 112 of `GeneralEnvelope.h`
* [x] Confirm a negative sign relating to dot products, line 78 of `JanusFactor.h`
* [x] Confirm math in `PairModulator.h`, line 218
* [x] Add `type_shapes` in `aniso.py`.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->

## How has this been tested?

I wrote expressions in Mathematica and had it do the math to check the values used in the potential. That notebook will be attached to this PR.

<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

<details>
<summary>Old nomenclature lookup table</summary>

Use this to translate when comparing old and new versions of the code.

<table id="orge7ef0a8" border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Kind of Entity</th>
<th scope="col" class="org-left">in Old Plugin</th>
<th scope="col" class="org-left">in New Code</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">thing created from template</td>
<td class="org-left">JanusDirectionalEvaluator</td>
<td class="org-left"><del>JanusEnvelope</del></td>
</tr>


<tr>
<td class="org-left"><b>Header file</b></td>
<td class="org-left">EvaluatorPairIsoModulated.h</td>
<td class="org-left">PairModulator.h</td>
</tr>


<tr>
<td class="org-left">template typename</td>
<td class="org-left">isoEval</td>
<td class="org-left">pairEvaluator</td>
</tr>


<tr>
<td class="org-left">variable</td>
<td class="org-left">iParam (&ldquo;isotropic&rdquo;)</td>
<td class="org-left">pairParam</td>
</tr>


<tr>
<td class="org-left">variable</td>
<td class="org-left">iP (for &ldquo;isotropic parameter&rdquo;), _iP</td>
<td class="org-left">pairP, _pairP</td>
</tr>


<tr>
<td class="org-left">template typename</td>
<td class="org-left">dirEval</td>
<td class="org-left">directionalEnvelope</td>
</tr>


<tr>
<td class="org-left">variable <b>?</b></td>
<td class="org-left">dParam</td>
<td class="org-left">envelopeParam</td>
</tr>


<tr>
<td class="org-left">variable</td>
<td class="org-left">dP, _dP</td>
<td class="org-left">envelP, _envelP</td>
</tr>


<tr>
<td class="org-left">protected variable</td>
<td class="org-left">iEv (&ldquo;isotropic (pair) evaluator&rdquo;)</td>
<td class="org-left">pairEval</td>
</tr>


<tr>
<td class="org-left">protected variable</td>
<td class="org-left">dEv (&ldquo;directional (pair) evaluator&rdquo;)</td>
<td class="org-left">envelEval</td>
</tr>


<tr>
<td class="org-left">variable in PairModulator</td>
<td class="org-left">isoModulator</td>
<td class="org-left">envelope</td>
</tr>


<tr>
<td class="org-left"><b>Header file</b></td>
<td class="org-left">DirectionalEvaluatorPairSphere.h</td>
<td class="org-left">GeneralEnvelope.h</td>
</tr>


<tr>
<td class="org-left">old header file with bugs</td>
<td class="org-left"><del>DirectionalEvaluatorPair</del></td>
<td class="org-left">&#xa0;</td>
</tr>


<tr>
<td class="org-left">returned name</td>
<td class="org-left">deps</td>
<td class="org-left">genenv</td>
</tr>


<tr>
<td class="org-left">template typename</td>
<td class="org-left">AngularModulator</td>
<td class="org-left">AngleDependence</td>
</tr>


<tr>
<td class="org-left"><b>Header file</b></td>
<td class="org-left">ModulatorJanus(&ldquo;Sphere&rdquo; in code).h</td>
<td class="org-left">JanusFactor.h</td>
</tr>


<tr>
<td class="org-left">Header file</td>
<td class="org-left">AllAnisoPairPotentialsExt.h</td>
<td class="org-left">Linking done in &ldquo;module.cc&rdquo; and &ldquo;module-md.cc&rdquo;</td>
</tr>


<tr>
<td class="org-left">exported name</td>
<td class="org-left">PotentialPairJanusLJ</td>
<td class="org-left">PotentialPairJanusLJ</td>
</tr>


<tr>
<td class="org-left">&#xa0;</td>
<td class="org-left">AnisoPotentialPair</td>
<td class="org-left">[already exists]</td>
</tr>


<tr>
<td class="org-left">Header file</td>
<td class="org-left">EvaluatorPairAnisoModulated.h</td>
<td class="org-left">?</td>
</tr>
</tbody>
</table>
</details>



## Change log

<!-- Propose a change log entry. -->
```

```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

cbkerr committed:
Indentation fixes

cbkerr committed:
Merge branch 'trunk-major' into feature/janus-pair

cbkerr committed:
Add AnisoPotentialPairJanus and friends copied from GB

cbkerr committed:
Fix typos

cbkerr committed:
Copy GB init code to aniso.py

cbkerr committed:
Typo fix

cbkerr committed:
Update class names for JanusLJ

cbkerr committed:
Update instantiating naming

cbkerr committed:
Add JanusFactor as copy of ModulatorJanus

cbkerr committed:
Add meat of GeneralEnvelope.h

Used to be called DirectionalEvaluatorPairSphere.h

cbkerr committed:
Add example of template instantiating JanusMie potential

cbkerr committed:
Add bones of PairModulator.h

Adapted from EvaluatorPairIsoModulated.h

cbkerr committed:
Add params to EvaluatorPairJanus.h even though file should be removed

This is what the template instantiated version of the new pair
potential should look like.

cbkerr committed:
What will get the new potential to compile

cbkerr committed:
Add body to PairModulator. See translation table in details.

| Name in Old Plugin                   | Name in New Code    |
+--------------------------------------+---------------------+
| EvaluatorPairIsoModulated.h          | PairModulator.h     |
| isoEval                              | pairEvaluator       |
| iParam ("isotropic")                 | pairParam           |
| iP (for "isotropic parameter"), _iP  | pairP               |
| dirEval                              | directionalEnvelope |
| dParam                               | envelopeParam       |
| dP, _dP                              | envelP              |
| iEv ("isotropic (pair) evaluator")   | pairEval            |
| dEv ("directional (pair) evaluator") | envelEval           |
| isoModulator                         | envelope            |

cbkerr committed:
Add docstrings

cbkerr committed:
Add docstring and name of GeneralEnvelope

cbkerr committed:
Fix typo

cbkerr committed:
Substitute old functions in JanusFactor

cbkerr committed:
update cmakelist

cbkerr committed:
Include headers in template instiantiating file

cbkerr committed:
Fix names in export function

cbkerr committed:
Update header guards and namespace of PairModulator

cbkerr committed:
Fix first rounds of compiler errors

cbkerr committed:
Define shape_type struct

cbkerr committed:
Define HOSTDEVICE

cbkerr committed:
Cast to right types for quat rotate

cbkerr committed:
Avoid intermediate JanusEnvelope creation

cbkerr committed:
Fix spacing in other files

cbkerr committed:
Add tags to demo potentials

cbkerr committed:
Add toPython method and py constructor for PairModulatorParamStruct

cbkerr committed:
Add changes from meeting Josh

cbkerr committed:
Fix constructor methods

cbkerr committed:
Replace toPython with as Dict

cbkerr committed:
Add param_type to GeneralEnvelope and JanusFactor

cbkerr committed:
Finish link to Python

cbkerr committed:
Improve understandability

cbkerr committed:
Fix bug with lack of cos(alpha), precomputing it

JanusFactor now matches equation (4) in Beltran-Villegas.

cbkerr committed:
Improve torque spacing

cbkerr committed:
Add input range validation and rename key to envelope_params

cbkerr committed:
Remove unused EvaluatorPairJanus.h

cbkerr committed:
Fix some comment syntax

cbkerr committed:
Remove todo note

cbkerr committed:
Convert some long form cross and dot to built in functions

cbkerr committed:
Fix some casts

cbkerr committed:
Address some todos

cbkerr committed:
Add documentation and move away from Janus naming

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

b-butler commented on hoomd/md/EvaluatorPairDipole.h:
```diff
@@ -110,7 +110,7 @@ class EvaluatorPairDipole
             }
 
-        pybind11::object toPython()
+        pybind11::object asDict()
```
We are returning a tuple here. That is why the method is named `toPython` because not every type parameter used here is a `dict` or mapping. Please revert the name change.

b-butler commented on hoomd/md/EvaluatorPairGB.h:
```diff
@@ -117,7 +117,7 @@ class EvaluatorPairGB
         shape_type(pybind11::object shape_params, bool managed) { }
 
-        pybind11::object toPython()
+        pybind11::object asDict()
```
Another example of not returning a dict.

b-butler commented on hoomd/md/pair/aniso.py:
```diff
@@ -574,3 +575,98 @@ def type_shapes(self):
+        * ``envelope_params`` (`dict`, **required**)
+
+            * ``alpha`` (`float`) - the patch half-angle :math:`\alpha`
+            * ``omega`` (`float`) - the patch steepness :math:`\omega`
```
I think having these as two seperate `TypeParameter`s would be preferred.

b-butler commented on hoomd/md/pair/aniso.py:
```diff
@@ -574,3 +575,98 @@ def type_shapes(self):
+                                 sigma = float),
+                envelope_params = dict(alpha = OnlyTypes(float,
+                                                        postprocess = self._check_0_pi),
+                                       omega= float)),
```
Normally we use the explicit `{"param": spec}` and I think that is cleaner for these also marginally faster.

b-butler commented on hoomd/md/pair/aniso.py:
```diff
@@ -574,3 +575,98 @@ def type_shapes(self):
+
+        Example:
+
+            >>> TODO
```
Pointing out existing todo.

b-butler commented on hoomd/md/EvaluatorExternalElectricField.h:
```diff
@@ -58,7 +58,7 @@ class EvaluatorExternalElectricField
+        pybind11::object asDict()
             {
             pybind11::tuple params;
             params = pybind11::make_tuple(E.x, E.y, E.z);
```
Just return the result of `make_tuple`.

b-butler commented on hoomd/md/AnisoPotentialPairJanusGPU.cc:
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
```
Is this filename right?

b-butler commented on hoomd/md/AnisoPotentialPairJanusGPUKernel.cu:
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
```
Is this filename right?

b-butler commented on hoomd/md/pair/aniso.py:
```diff
@@ -574,3 +575,98 @@ def type_shapes(self):
         return self._return_type_shapes()
+
+
+class JanusLJ(AnisotropicPair):
```
Where is the `Mie` potential one?

b-butler commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,143 @@
+    DEVICE GeneralEnvelope(pybind11::dict params)
+        : param_type(params)
+        {
+        }
```
This explicitly can't be compiled on the DEVICE though.

b-butler commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,143 @@
+            Scalar jPi = modPj*modi/s.magdr;
+            
+            // torque on ith
+            torque_i = vec_to_scalar3(iPj * cross(vec3<Scalar>(s.ei), vec3<Scalar>(s.dr))); // TODO: is all the casting efficient?
```
Copying a `Scalar3` shouldn't be too costly.

b-butler commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,270 @@
+                }
+
+            // compute envelope
+            Scalar envelope(Scalar(0));
```
```suggestion
            Scalar envelope(0.0);
```

b-butler commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,270 @@
+            envelEval.evaluate(force, envelope, torque_i, torque_j);
+
+            // modulate forces
+            // TODO check this math
```
Did you?

b-butler commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,270 @@
+            force.z = pair_eng*force.z + dr.z*force_divr*envelope;
+
+            // modulate torques
+            // TODO check this math
```
Did you?

b-butler commented on hoomd/md/JanusFactor.h:
```diff
@@ -0,0 +1,126 @@
+
+            // compute dot products
+            doti = dot(vec3<Scalar>(dr), ei) / magdr; // cos(angle between dr and ei)
+            // doti = -(dr.x*ei.x+dr.y*ei.y+dr.z*ei.z)/magdr; // TODO why was this negative?
```
Do you know now?

b-butler commented on hoomd/md/JanusFactor.h:
```diff
@@ -0,0 +1,126 @@
+        }
+
+
+    // things that get passed in to constructor
```
These comments are unnecessary. Discussing what they are is more important.

b-butler commented on hoomd/md/JanusFactor.h:
```diff
@@ -0,0 +1,126 @@
+    {
+/*! \file JanusFactor.h
+*/
+class JanusFactor
```
It would be helpful to document this class.

b-butler commented on hoomd/md/JanusFactor.h:
```diff
@@ -0,0 +1,126 @@
+        : dr(_dr), qi(_qi), qj(_qj), params(_params)
+        {
+            // compute current janus direction vectors
+            vec3<Scalar> e { make_scalar3(1, 0, 0) };
```
So did we decide that expanding this to allow for multiple patches was out of bounds for this PR?

b-butler reviewed (dismissed):
Left a number of comments to address. Thanks for the hard work @cbkerr.

Commit:
changing Janus GPU files to right templates

cbkerr committed:
Reverse format scrunch

cbkerr committed:
Confirm sign of original code

cbkerr committed:
Add in progress note about torque derivation

cbkerr committed:
Add new general torque multiplier for particle i

cbkerr committed:
Add more reference rotation vectors for clarity during derivation

cbkerr committed:
Address some compilation errors

cbkerr committed:
Try adding patch orientation to JanusFactor

cbkerr committed:
Add other vector types

cbkerr committed:
Add notes about how to pass in per-type orientation of patch

cbkerr committed:
Clean up initializing variables

cbkerr committed:
Add possible places to add patch orientation

cbkerr committed:
Add notes from meeting with Josh

cbkerr committed:
Use quat instead of Scalar4

cbkerr committed:
Merge JanusFactor into GeneralEnvelope

cbkerr committed:
Remove JanusFactor

cbkerr committed:
Remove incorrect ways to store patch direction

cbkerr committed:
Finish merging into GeneralEnvelope

cbkerr committed:
Convert given patch directions to quats

cbkerr committed:
Fix casting tuple bug

cbkerr committed:
Switch to explicit dict syntax

cbkerr committed:
Apply rest of conj(quat) bug

cbkerr committed:
Add math discrep note

cbkerr committed:
In progress torque bug fix

cbkerr committed:
Debugging work from mid January

cbkerr committed:
Notes from easier derivation a la Philipp

cbkerr committed:
sort cmakelist of tests

cbkerr committed:
Skeleton test code

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Add rest of test params

cbkerr committed:
Fix test structure

cbkerr committed:
Use params.ni to avoid not initializing

cbkerr committed:
Remove extraneous and debugging comments

cbkerr committed:
Label todos in testing

cbkerr committed:
Fix bug Philipp found

cbkerr committed:
Update tests

cbkerr committed:
Update force values after adding missing dr/magdr factor to test

cbkerr committed:
Add tolerance options and update negation in tests

cbkerr committed:
Check that we get normalized ni, nj back

cbkerr committed:
Add more test cases from Mathematica that vary n and alpha

cbkerr committed:
Update test check

cbkerr committed:
Scan negative signs with Philipp

cbkerr committed:
Add PatchyYukawa

cbkerr committed:
Add python test generation code

cbkerr committed:
Fix tests and generation

cbkerr committed:
My implementation following the math better

cbkerr committed:
Quaternion rotates particle (1,0,0) etc to world. No conj needed

cbkerr committed:
Remove print statements

cbkerr committed:
Add missing factor of 1/magdr

cbkerr committed:
Offline code review with Philipp

cbkerr committed:
replace df_dn with df_du to factor out rhat

cbkerr committed:
use my method bc other has wrong y component

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Add notes comparing force term versions

cbkerr committed:
Rectify difference between derivation methods

cbkerr committed:
Remove printing of force

cbkerr committed:
Enable PatchyExpandedGaussian

cbkerr committed:
Add groundwork for list of patches but only assign to first element

cbkerr committed:
Apply corrections to patch generation code

cbkerr committed:
Test params from 20ish days ago

cbkerr committed:
Get params from Python, use shape_type for patch properties

cbkerr committed:
Work with Brandon to make patch a particle property

joaander committed:
Merge branch 'trunk-minor' into feature/janus-pair

joaander committed:
Reapply removal of test_meshbond.py.

joaander committed:
Compile without error on Linux (CPU).

joaander committed:
Revert some unnecessary changes.

joaander committed:
Fix duplicate lines.

joaander committed:
Parse only one list of n vectors.

joaander committed:
Fix compile errors.

joaander committed:
Fix setting of patch directions.

joaander commented:
@cbkerr continues to work on this pull request.

cbkerr committed:
Remove extra magdr in testing code

cbkerr committed:
Make forcedivr more clear in testing code

cbkerr committed:
Remove run on main

cbkerr committed:
Add new_print_values to compare with new API

cbkerr committed:
Remove old patch test parameters

cbkerr committed:
Update tests to support new API for patches

cbkerr committed:
Comment out other Patchy potentials for now

cbkerr committed:
Update test_patch

cbkerr commented:
I'm still working on validating the potential

cbkerr committed:
Enable PatchyYukawa with new API

cbkerr committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/janus-pair

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Fix GPU compilation

cbkerr committed:
TODO: fix iteration over patch arrays of different lengths

If particle i has more patches than particle j, combinations of
patches where patchi index > patchj index are not calculated.

cbkerr committed:
Note 3 TODOs for performance

cbkerr committed:
Use rotmat3 in all cases

Improves TPS from 12.5 to 15.7.
Might need to enable different cases for __HIPCC__ or not in future.

cbkerr committed:
Remove pass-through functions Modulatori and Modulatorj

cbkerr committed:
Only compute fi and fj once and pass into derivatives

cbkerr committed:
Precompute exponents instead of costhetas

joaander commented:
Not stale.

cbkerr committed:
Some wording cleanups

cbkerr committed:
Working on GPU support

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Latest work towards GPU compilation

cbkerr committed:
Cleanup code and comments

cbkerr committed:
Switch to normalized modulator and move modulator functions inline

cbkerr committed:
Switch back to p for patch vector in particle frame, clarify naming

cbkerr committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/janus-pair

cbkerr committed:
Add some implemented torque tests

cbkerr committed:
Begin adding templates

cbkerr committed:
Fix CPU templates and add GPU templates

cbkerr committed:
Fix initialization order warnings

cbkerr committed:
Fix cpp links in Python

cbkerr committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/janus-pair

cbkerr committed:
Fix GPU build

cbkerr committed:
Add Python bindings

cbkerr committed:
Fix GPU build and ExpandedMie bug

cbkerr committed:
Prototype template-based docstrings

cbkerr committed:
Document rest of patchy potentials

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Swap negative signs for one fewer

cbkerr committed:
Fix docs

cbkerr committed:
Add some test cases

cbkerr committed:
Document normalization math

cbkerr committed:
Loop over all patch combinations

cbkerr committed:
Update patchy doc

cbkerr committed:
Update doc headers and examples

cbkerr committed:
Fix doc bug

cbkerr committed:
Enable setting different patch lists to A and B type in test

cbkerr committed:
Start writing multi patch docs

cbkerr committed:
Add tests from mathematica

cbkerr committed:
remove some comments

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

cbkerr committed:
Fix normalization checking

cbkerr committed:
Add all tests

joaander commented on hoomd/md/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ set(_md_sources module-md.cc
                    AnisoPotentialPairALJ3.cc
                    AnisoPotentialPairDipole.cc
                    AnisoPotentialPairGB.cc
+                   # AnisoPotentialPairJanus.cc
```
```suggestion
```

joaander commented on hoomd/md/CMakeLists.txt:
```diff
@@ -157,6 +158,7 @@ set(_md_headers ActiveForceComputeGPU.h
                 NeighborListTree.h
                 OPLSDihedralForceComputeGPU.h
                 OPLSDihedralForceCompute.h
+                PairModulator.h
```
Also add `GeneralEnvelope.h` above (in alphabetical order).

joaander commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,310 @@
+  The GeneralEnvelope creates the pair potential modulator.
+*/
+
+        static inline std::string vecString(vec3<Scalar> a) {
+            return std::to_string(a.x) + ", " + std::to_string(a.y) + ", " + std::to_string(a.z) + '\n';
+        }
+
```
```suggestion
```

joaander commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,310 @@
+namespace md
+    {
+
+/*
+  The GeneralEnvelope creates the pair potential modulator.
+*/
```
Move this down to right above the `GeneralEnveolope` class. Make it a [doxygen style comment](https://hoomd-blue.readthedocs.io/en/v4.8.2/style.html#id3). Expand the description to explain the general functional form of an envelope and explain how this class interoperates with the other classes.

joaander commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,310 @@
+        Scalar3 m_norm_patch_local_dir;
+    };
+
+    DEVICE GeneralEnvelope( // TODO replace GeneralEnvelope with PatchesEnvelope
```
Add doxygen style comments and describe all parameters.

joaander commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,310 @@
+        // TODO decide which is faster on GPU and CPU. // use 250_000 particles to saturate GPU // benchmark on brett
+        // : dr(_dr), qi(_quat_i), qj(_quat_j), params(_params), p_i(shape_i.m_norm_patch_local_dir), p_j(shape_j.m_norm_patch_local_dir)
+// #else
+        : dr(_dr), R_i(rotmat3<ShortReal>((quat<ShortReal>)_quat_i)), R_j(rotmat3<ShortReal>((quat<ShortReal>)_quat_j)), params(_params), p_i(shape_i.m_norm_patch_local_dir), p_j(shape_j.m_norm_patch_local_dir)
```
Unless you have benchmarks that show otherwise, I think all the `ShortReals` should be `LongReal`. In most cases, converting back and forth between `float` and `double` is far more expensive than the cost savings of a few math operations in `float`.

joaander commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,310 @@
+    vec3<Scalar> b1, b2, b3;
+
+
+    Scalar drsq;
+    Scalar magdr;
+    vec3<Scalar> rhat;
+
+    Scalar exp_negOmega_times_CosThetaI_minus_CosAlpha;
+    Scalar exp_negOmega_times_CosThetaJ_minus_CosAlpha;
```
HOOMD uses the `lower_case_with_underscores` naming convention for variables.

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+
+/*
+  This class aplies the DirectionalEnvelope to the PairEvaluator, turning the isotropic pair potential into an anisotropic potential.
+*/
```
Move this down to right above the `PairModulator` class. Make it a [doxygen style comment](https://hoomd-blue.readthedocs.io/en/v4.8.2/style.html#id3). Expand the description to explain the general functional form of the modulator and explain how this class interoperates with the other classes.

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+            {
+            }
+
+        param_type(typename PairEvaluator::param_type _pairP, typename DirectionalEnvelope::param_type _envelP)
+            {
+                pairP = _pairP;
+                envelP = _envelP;
+            }
```
```suggestion
```
I don't think this is used.

joaander commented on hoomd/md/GeneralEnvelope.h:
```diff
@@ -0,0 +1,310 @@
+      \param dj Diameter of particle j
+    */
+    DEVICE void setDiameter(Scalar di, Scalar dj) { }
+
```
```suggestion
```
These API calls have been removed.

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+                          const Scalar _rcutsq,
+                          const param_type& _params)
+        : dr(_dr),
+          rsq(_dr.x*_dr.x + _dr.y*_dr.y + _dr.z*_dr.z),
```
```suggestion
          rsq(dot(_dr, _dr)),
```

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+    DEVICE void setDiameter(Scalar di, Scalar dj)
+        {
+        }
+
```
```suggestion
```

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+            // if (PairEvaluator::needsCharge())
+            //     pairEval.setCharge(qi, qj);
+            // if (DirectionalEnvelope::needsCharge())
+            //     envelEval.setCharge(qi, qj);
```
```suggestion
        m_charge_i = qi;
        m_charge_j = qj;
```
Need to make member variables that store the charges and tags. A followup comment in `evaluate` uses these.

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+    //! Whether the pair potential needs particle tags.
+    HOSTDEVICE static bool needsTags()
+        {
+            return false;
```
```suggestion
        return (PairEvaluator::needsTags() || DirectionalEnvelope::needsTags());
```

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+    //! No modulated potential needs tags
+    HOSTDEVICE void setTags(unsigned int tagi, unsigned int tagj)
+        {
+        }
```
```suggestion
        {
        m_tag_i = tagi;
        m_tag_j = tagj;        
        }
```

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,351 @@
+                                }
+
+                            DirectionalEnvelope envel_eval(dr, quat_i, quat_j, rcutsq, params.envelP, shape_i->envelope[patchi], shape_j->envelope[patchj]);
+
```
Add code similar to the above to set charges and tags on the `envel_eval`.

joaander reviewed (changes requested):
Thanks! This has come a long way from where we started. The final implementation is very clean and easy to follow. I have some minor suggestions for clean up and code style consistency. Let's get this wrapped up.

cbkerr committed:
Apply quick suggestions from code review

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

cbkerr committed:
Member variables for charges and tags

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

cbkerr committed:
Implement changes from code review with Josh

cbkerr committed:
remove old in-progress comments

cbkerr committed:
Update doxygen comments

cbkerr committed:
remove type_shapes logger

cbkerr committed:
updating comments

cbkerr committed:
Fix compile bug

cbkerr committed:
Rename GeneralEnvelope to PatchEnvelope

cbkerr committed:
Update PatchEnvelope

cbkerr committed:
Clean up PairModulator

cbkerr committed:
Clean up PatchEnvelope

cbkerr committed:
Remove generate_test_patch

cbkerr committed:
Run precommit

cbkerr committed:
Placate the formatter

cbkerr committed:
Start formatting aniso.py

joaander committed:
Fix compile errors and warnings.

cbkerr committed:
Remove untested PatchyTable class

cbkerr committed:
change line length

cbkerr committed:
Format aniso module docstrings

cbkerr committed:
Add sybil example to Patchy class

cbkerr committed:
Update Patchy examples for sybil and move examples above attributes

cbkerr committed:
Placate flake8

cbkerr committed:
Merge branch 'trunk-minor' into feature/janus-pair

joaander committed:
Fix https link.

cbkerr commented:
This is the test generation file for Mathematica. I added the extra .txt because github doesn't like .nb files.
[patchy-tests-committed.nb.txt](https://github.com/user-attachments/files/17263237/patchy-tests-committed.nb.txt)

joaander commented on hoomd/md/PairModulator.h:
```diff
@@ -0,0 +1,307 @@
+        {
+        param_type() { }
+
+        // param_type(typename PairEvaluator::param_type _pair_p, typename
+        // DirectionalEnvelope::param_type _envel_p)
+        //     {
+        //         pair_p = _pair_p;
+        //         envel_p = _envel_p;
+        //     }
```
```suggestion
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+        patchy.patches.default = []
+        simulation.operations.integrator.forces = [patchy]
+
+    For multiple patches, every combination of patch and
+    We use PatchyLJ as the example.
+
+    To make specific patches
+
+    todo write example
```
```suggestion
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyLJ(Patchy):
+    """Directional version of LJ."""
```
```suggestion
    """Modulate `hoomd.md.pair.LJ` with angular patches."""
```

tcmoore3 replied:
+1

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyExpandedGaussian(Patchy):
+    """Directional version of ExpandedGaussian."""
```
```suggestion
    """Modulate `hoomd.md.pair.ExpandedGaussian` with angular patches."""
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyExpandedLJ(Patchy):
+    """Directional version of ExpandedLJ."""
```
```suggestion
    """Modulate `hoomd.md.pair.ExpandedLJ` with angular patches."""
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyExpandedMie(Patchy):
+    """Directional version of ExpandedMie."""
```
```suggestion
    """Modulate `hoomd.md.pair.ExpandedMie` with angular patches."""
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyGaussian(Patchy):
+    """Directional version of Gaussian."""
```
```suggestion
    """Modulate `hoomd.md.pair.Gaussian` with angular patches."""
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyMie(Patchy):
+    """Directional version of Mie."""
```
```suggestion
    """Modulate `hoomd.md.pair.Mie` with angular patches."""
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+
+class PatchyYukawa(Patchy):
+    """Directional version of Yukawa."""
```
```suggestion
    """Modulate `hoomd.md.pair.Yukawa` with angular patches."""
```

joaander commented:
@DomFijan let us know if you have any comments on the documentation: https://hoomd-blue--1412.org.readthedocs.build/en/1412/module-md-pair-aniso.html#hoomd.md.pair.aniso.Patchy

DomFijan commented:
I can take a look at this Wednesday/Thursday.

DomFijan commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+        .. code-block:: python
+
+            patchy.patches['A'] = [(1,0,0), (1,1,1)]
+    """
```
"mode" documentation is missing. It's missing for all classes in this file that have been added.

joaander replied:
There are many attributes and methods that HOOMD does not duplicate the documentation of. The assumption is that you read the base class documentation to learn about common methods: https://hoomd-blue--1412.org.readthedocs.build/en/1412/module-md-pair.html#hoomd.md.pair.Pair.mode

Should we choose to change that, we can instruct sphinx to repeat inherited docs - although I don't know how this handles attributes that we explicitly document with `.. py::attribute`.

joaander replied:
I opened #1914 as a proposal to repeat abbreviated docs for all inherited attributes and methods. We can follow the proposed pattern for `Patchy` in this PR and then later extend it to the rest of HOOMD-blue in future work.

DomFijan commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+    .. _Beltran-Villegas et. al.: https://dx.doi.org/10.1039/C3SM53136H
+
+    .. math::
+
+        U(r, \theta_i, \theta_j) = f(\theta_i, \alpha, \omega)
+        f(\theta_j, \alpha, \omega)
+        U_{\mathrm{pair}}(r)
+
+    where :math:`f` is an orientation-dependent factor of the patchy spherical
+    cap half-angle :math:`\alpha` and patch steepness :math:`\omega`.
+
+    .. math::
+        \begin{align}
+        f(\theta, \alpha, \omega) &= \frac{\big(1+e^{-\omega (\cos{\theta} -
+        \cos{\alpha}) }\big)^{-1} - f_{min}}{f_{max} - f_{min}}\\
+        f_{max} &= \big( 1 + e^{-\omega (1 - \cos{\alpha}) } \big)^{-1} \\
+        f_{min} &= \big( 1 + e^{-\omega (-1 - \cos{\alpha}) } \big)^{-1} \\
+        \end{align}
+
+    .. invisible-code-block: python
+        patchy = hoomd.md.pair.aniso.PatchyLJ(nlist = neighbor_list,
+                                              default_r_cut = 3.0)
+        pair_params = {'epsilon': 10, 'sigma': 1}
+
+        patchy.params.default = dict(pair_params = pair_params,
+                                     envelope_params = {'alpha': math.pi/4,
+                                                        'omega': 30})
+        patchy.patches.default = []
+        simulation.operations.integrator.forces = [patchy]
```
I know we want to avoid having images/figures in docs, but a figure here plotting envelope as fn of alpha and omega would really help. Another issue is that omega is not bounded on the upper side so it might be difficult to guess what a reasonable value is. Alternatively, perhaps you could add a sentence explaining what a reasonable choice for omega would be (or a range). I think in general alpha is pretty self explanatory.

joaander replied:
svg figures are welcome in the docs, specially when needed to add clarity like this. Add a notebook to `sphinx-doc/figures` that generates the figure - follow the same style as the notebooks there. Please don't commit the executed notebook cells or else we will need to squash merge this commit.

DomFijan commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+"""
+
+
+class Patchy(AnisotropicPair):
+    r"""Pair potential made directional by an envelope.
```
Parameters are not listed? Args section is missing I think...

joaander replied:
The args section is intentionally left missing in the base class. The user should not instantiate it.

DomFijan replied:
I'm looking at another base class `hoomd.md.pair.aniso.AnisotropicPair` and that one does have Args section listed on the docs, but some other such  as `hoomd.md.pair.Pair` does not...

DomFijan replied:
If the user shouldn't instantiate this class, we should have a note stating that this class should never be instantiated.

joaander replied:
> I'm looking at another base class `hoomd.md.pair.aniso.AnisotropicPair` and that one does have Args section listed on the docs, but some other such as `hoomd.md.pair.Pair` does not...

OK, I guess we are not consistent on this. I was thinking of other base classes like:
* https://hoomd-blue.readthedocs.io/en/v4.8.2/module-md-pair.html#hoomd.md.pair.Pair
* https://hoomd-blue.readthedocs.io/en/v4.8.2/module-md-methods.html#hoomd.md.methods.Method
* https://hoomd-blue.readthedocs.io/en/v4.8.2/module-md-force.html#hoomd.md.force.Force
* etc...

> If the user shouldn't instantiate this class, we should have a note stating that this class should never be instantiated.

Yes, we should add this note.

DomFijan commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+    angular patches with an implementation based on
+    `Beltran-Villegas et. al.`_. With distance between particles
+    :math:`r` and minimal angles between respective particle direction and
+    the displacement vector :math:`\theta_i`, :math:`\theta_j` the patchy potential
```
Are theta's here angles between particle orientation and patch direction for two interacting particles? In any case this should be better explained. Maybe you can also consider using notation compatible with Kern-Frenkel model from hoomd tutorial?

tcmoore3 replied:
Agreed. Leaving a suggestion

DomFijan commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+    .. py:attribute:: patches
+
+        List of vectors pointing to patch centers. Unnormalized vectors are normalized.
+        If a particle type does not have patches, set the patches to an empty list.
```
If a particle does not have patches is the whole particle sticky?

joaander replied:
No, then `Pathcy*` provides no interaction at all. You also need to apply `pair.LJ` to make the whole particle volume excluding (or sticky if needed).

We should clarify this by writing the patch interaction as a double sum over the patches. When either sum is over 0 patches, then the whole sum is 0.

DomFijan replied:
Should we even allow users to make another particle patchless? What would be the purpose of that since they are not interacting at all in that case?

joaander replied:
The assumption is that you use a volume-excluding potential in addition to a `Pathcy` one. Otherwise, even your patch particles can pass through each other when the patches are _not_ aligned.

One example where users may want no patches: Explicit solvent simulations with an isotropic solvent and patchy particles:
```
wca = hoomd.md.pair.LJ(nlist=..., default_r_cut=2**(1.0/6.0))
wca.params.default= {'sigma': 1.0, 'epsilon': 1.0}

patchy_lj = hoomd.md.pair.ansio.PatchyLJ(nlist=..., default_r_cut=2.5)
patchylj.params.default = dict(pair_params = {'sigma': 1.0, 'epsilon': 1.0},
                               envelope_params = envelope_params)

patchy_lj.patches['A'] = [(1,0,0), (-1,0,0)]
patchy_lj.patches['solvent'] = []
```

Here `wca` provides the hard core repulsion. `patchy_lj` adds attractive patches on top of that for _A-A_ particle interactions. The `patchy_lj` energy is 0 for _A-solvent_ and _solvent-solvent_ particle pairs.

Edit: While having no patches is _sufficient_ to set the patch energy to 0, you can make the calculation more efficient with:
```
patchy_lj.r_cut[('A', 'solvent')] = 0
patchy_lj.r_cut[('solvent', 'solvent')] = 0
```

DomFijan replied:
This should probably be documented?

tcmoore3 replied:
I don't think it belongs in the reference documentation., but it could make a nice how-to guide.

joaander replied:
This would be good material for an eventual tutorial.

DomFijan reviewed (changes requested):
I have a couple of questions and suggestions for improvements on the docs side.

tcmoore3 commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+        Type: `TypeParameter` [`tuple` [``particle_type``, ``particle_type``],
+        `dict`]
+
+    .. py:attribute:: patches
```
This parameter name is too vague. More specific options, from my favorite to least favorite:
1. `directors`
2. `patch_directors`
3. `patch_directions`
4. `patch_vectors`
5. `patch_direction_vectors`.

tcmoore3 commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+    .. py:attribute:: params
+
+        The Patchy potential parameters. A dictionary of dictionaries that
+        are passed on to the envelope and to the pair potential.
+
+        The dictionary has the following keys:
```
```suggestion
        The Patchy potential parameters. The dictionary has the following keys:
```

tcmoore3 commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+    .. code-block:: python
+
+        lj_params = dict(epsilon = 1, sigma = 1)
```
```suggestion
        lj_params = dict(epsilon=1, sigma=1)
```
...and in the other code examples for consistency with the rest of the docs.

tcmoore3 commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+    .. code-block:: python
+
+        expanded_mie_params = {'epsilon': 1, 'sigma': 1,
+                               'n': 10, 'm': 15, 'delta': 1}
```
```suggestion
                               'n': 15, 'm': 10, 'delta': 1}
```

tcmoore3 commented on hoomd/md/pair/aniso.py:
```diff
@@ -563,3 +572,425 @@ def type_shapes(self):
+
+    .. code-block:: python
+
+        mie_params = {'epsilon': 1, 'sigma': 1, 'n': 10, 'm': 15}
```
```suggestion
        mie_params = {'epsilon': 1, 'sigma': 1, 'n': 15, 'm': 10}
```

tcmoore3 reviewed (approved):
I went through the docs and made a few suggestions and mostly agreed with what Josh and Dom have already commented.

The thought crossed my mind to separate these out into a separate `patchy` namespace, e.g., either `md.pair.aniso.patchy` or `md.pair.patchy`, mainly to keep the documentation of these new classes out of the way of the existing `md.pair.aniso` classes. Since `md.pair.aniso` is the most accurate description of all of these classes, it's probably best to keep it the way you have it.

cbkerr committed:
Apply cleanups from code review

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>
Co-authored-by: Tim Moore <mtimc@umich.edu>

cbkerr committed:
Update docstrings for patchy potential

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

cbkerr committed:
Add warning to not instantiate Patchy directly

joaander committed:
Fix errors in AngularStep documentation.

joaander committed:
Reword Patchy documentation.

joaander committed:
Merge branch 'trunk-minor' into feature/janus-pair

joaander committed:
Unified solution to inherited docs.

Also fully document the mathematical representation of the patchy
potential.

joaander committed:
patches->directors

joaander committed:
Merge branch 'feature/janus-pair' of github.com:glotzerlab/hoomd-blue into feature/janus-pair

cbkerr committed:
Add blanked notebook to generate patchy figures

cbkerr committed:
Add figures to docs with alt text

cbkerr committed:
Fix alt text indentation

cbkerr committed:
Apply pre-commit

cbkerr committed:
Remove spaces around = in parameter lists and swap n and m in Mie

cbkerr commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

joaander committed:
Fix pre-commit errors.

joaander committed:
Fix definition of patchy potential.

joaander reviewed (approved):
THANKS!
